sw/device/tests/crypto/cryptotest/firmware/ecdsa.c

```diff
@@ -0,0 +1,187 @@
+      public_key.key_length = sizeof(p256_point_t);
+      public_key.key = (uint32_t *)&pub_p256;
+      digest_len = kP256ScalarWords;
+      if (uj_signature.r_len > kP256ScalarBytes) {
+        LOG_ERROR(
+            "Signature r value too large for P256 (have = %d bytes, max = %d "
+            "bytes)",
+            uj_signature.r_len, kP256ScalarBytes);
+        return INVALID_ARGUMENT();
+      }
+      if (uj_signature.s_len > kP256ScalarBytes) {
+        LOG_ERROR(
+            "Signature s value too large for P256 (have = %d bytes, max = %d "
+            "bytes)",
+            uj_signature.s_len, kP256ScalarBytes);
+        return INVALID_ARGUMENT();
+      }
+      memset(signature_p256.r, 0, kP256ScalarBytes);
+      memcpy(signature_p256.r, uj_signature.r, uj_signature.r_len);
+      memset(signature_p256.s, 0, kP256ScalarBytes);
+      memcpy(signature_p256.s, uj_signature.s, uj_signature.s_len);
+      signature_mut.len = kP256ScalarWords * 2;
+      signature_mut.data = (uint32_t *)&signature_p256;
+      break;
+    default:
+      LOG_ERROR("Unsupported ECC curve: %d", uj_curve);
+      return INVALID_ARGUMENT();
+  }
+  public_key.checksum = integrity_unblinded_checksum(&public_key);
+  otcrypto_const_word32_buf_t signature = {
+      .len = signature_mut.len,
+      .data = signature_mut.data,
+  };
+
+  otcrypto_ecc_curve_t elliptic_curve = {
+      .curve_type = curve_type,
+      // NULL because we use a named curve
+      .domain_parameter = NULL,
+  };
+  otcrypto_hash_mode_t mode;
+  switch (uj_hash_alg) {
+    case kCryptotestEcdsaHashAlgSha256:
+      mode = kOtcryptoHashModeSha256;
+      break;
+    case kCryptotestEcdsaHashAlgSha384:
+      mode = kOtcryptoHashModeSha384;
+      break;
+    case kCryptotestEcdsaHashAlgSha512:
+      mode = kOtcryptoHashModeSha512;
+      break;
+    case kCryptotestEcdsaHashAlgSha3_256:
+      mode = kOtcryptoHashModeSha3_256;
+      break;
+    case kCryptotestEcdsaHashAlgSha3_384:
+      mode = kOtcryptoHashModeSha3_384;
+      break;
+    case kCryptotestEcdsaHashAlgSha3_512:
+      mode = kOtcryptoHashModeSha3_512;
+      break;
+    default:
+      LOG_ERROR("Unrecognized ECDSA hash mode: %d", uj_hash_alg);
+      return INVALID_ARGUMENT();
+  }
+  uint8_t message_buf[ECDSA_CMD_MAX_MESSAGE_BYTES];
+  memset(message_buf, 0, digest_len * sizeof(uint32_t));
+  memcpy(message_buf, uj_message.input, uj_message.input_len);
+  const otcrypto_hash_digest_t message_digest = {
+      .mode = mode,
+      .len = digest_len,
+      .data = (uint32_t *)message_buf,
+  };
+  switch (uj_op) {
+    case kCryptotestEcdsaOperationVerify: {
+      hardened_bool_t verification_result = kHardenedBoolFalse;
+      otcrypto_status_t status =
+          otcrypto_ecdsa_verify(&public_key, message_digest, signature,
+                                &elliptic_curve, &verification_result);
+      cryptotest_ecdsa_verify_output_t uj_output;
+      switch (status.value) {
```

Comment:
This section is starting to get hard to read with the switch cases nested so deeply. Can you pull all this status checking code out into a separate function?

Reply:
Split into separate functions